drivers/timer/cortex_m_systick.c

```diff
@@ -217,9 +221,27 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 			last_load = delay;
 		}
 	}
+
+	val2 = SysTick->VAL;
+
 	SysTick->LOAD = last_load - 1;
 	SysTick->VAL = 0; /* resets timer to last_load */
 
+	/*
+	 * Add elapsed cycles while computing the new load to cycle_count.
+	 *
+	 * Note that comparing val1 and val2 is normaly not good enough to
+	 * guess if the counter wrapped during this interval. Indeed if val1 is
+	 * close to LOAD, then there are little chances to catch val2 between
+	 * val1 and LOAD after a wrap. COUNTFLAG should be checked in addition.
+	 * But since the load computation is faster than MIN_DELAY, then we
+	 * don't need to worry about this case.
+	 */
+	if (val1 < val2) {
+		cycle_count += (val1 + (last_load_ - val2));
```

Comment:
Hi @semihalf-niedzwiecki-dawid,

`elapsed()` won't detect the overflow because writing 0 in the SysTick Current Value Register from `sys_clock_set_timeout()` clears the `COUNTFLAG` status bit to 0:

`SysTick->VAL = 0; /* resets timer to last_load */`

Reply:
Thanks for the response :-)

But what happens if the overflow occurs between `val1 = SysTick->VAL;` and `val2 = SysTick->VAL;`? Isn't `last_load` added to `overflow_cyc` and `val2 > val1` so the `cycle_count += (val1 + (last_load_ - val2));` is executed?

Reply:
No it is not. `sys_clock_set_timeout()` computes the new timer load with interrupts disabled. See the call to `k_spin_lock()`. So in case of overflow `sys_clock_isr()` will be called right after sys_clock_set_timeout() returned and will deal with the updated timer values and a `COUNTFLAG` status bit set to 0. There is a comment explaining that in `sys_clock_isr()`.

At the time I wrote this patch I tested this code a lot. Are you experimenting an issue ? Note that this driver is still leaking few cycles.

Reply:
>At the time I wrote this patch I tested this code a lot. Are you experimenting an issue ?
No, I'm just trying to fully understand the code, I haven't encountered the described case.

 >No it is not. sys_clock_set_timeout() computes the new timer load with interrupts disabled. See the call to k_spin_lock(). So in case of overflow sys_clock_isr() will be called right after sys_clock_set_timeout() returned and will deal with the updated timer values and a COUNTFLAG status bit set to 0. 
Thanks a lot for the explanation. I thought for some reason that the SysTick IRQ is not disabled by the `arch_irq_lock` call (as SVC), it was my mistake. Now everything is clear.

>Note that this driver is still leaking few cycles.
I guess ticks between `val2 = SysTick->VAL;` and `SysTick->VAL = 0;` are missed. It looks like the run time measurements and correcting `cycle_count` would make sense here.

Reply:
> I guess ticks between `val2 = SysTick->VAL;` and `SysTick->VAL = 0;` are missed. It looks like the run time measurements and correcting `cycle_count` would make sense here.

Yes. There are also a few ticks lost by hardware when reloading and resetting the timer. They all add up and make a drift which can become "not negligible" especially when high frequency timers are used by the application.

Reply:
Ok, thanks for the explanation again.